### PR TITLE
KAFKA-2856: add ktable

### PR DESCRIPTION
**Pull request by ymatsuda**
@guozhangwang 
- added KTable API and impl
- added standby support for KTable

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java [hunk 1]
```diff
@@ -65,6 +69,43 @@ public KStreamBuilder() {
     }
 
     /**
+     * Creates a KStream instance for the specified topic.
```
Comment (wushujames) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java]: "Creates a KTable instance"?

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java [hunk 2]
```diff
@@ -65,6 +69,43 @@ public KStreamBuilder() {
...
+     *
+     * @param topic          the topic name
+     * @return KTable
+     */
+    public <K, V> KTable<K, V> table(String topic) {
+        return table(null, null, null, null, topic);
+    }
+
+    /**
+     * Creates a KStream instance for the specified topic.
```
Comment (wushujames) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java]: "Creates a KTable instance"?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableSource.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        materialized = true;
+    }
+
+    public boolean isMaterialized() {
+        return materialized;
+    }
+
+    @Override
+    public KTableValueGetterSupplier<K, V> view(KTableValueGetterSupplier<K, V> valueGetterSupplier) {
+        throw new IllegalStateException();
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableSource.java]: nit: Add the error message into the exception?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.KStreamBuilder;
+import org.apache.kafka.streams.kstream.KTable;
+import org.apache.kafka.streams.kstream.Predicate;
+import org.apache.kafka.streams.kstream.ValueMapper;
+import org.apache.kafka.streams.processor.StateStoreSupplier;
+
+import java.util.Collections;
+
+public class KTableImpl<K, S, V> implements KTable<K, V> {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:37]: nit: adding some comments explaining generic type S?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    @Override
+    public KStream<K, V> toStream() {
+        String name = topology.newName(TOSTREAM_NAME);
+
+        topology.addProcessor(name, new KStreamPassThrough(), this.name);
+
+        return new KStreamImpl<>(topology, name, Collections.singleton(sourceNode));
+    }
+
+    KTableValueGetterSupplier<K, V> valueGetterSupplier() {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:171]: Is this function only going to be used for tests? If yes we could add some comments stating that.

EDIT: just realized it is going to be used for stateful operations which are not there yet, never mind.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableStoreSupplier.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        this.serdes = new Serdes<>(name, keySerializer, keyDeserializer, valSerializer, valDeserializer);
+        this.time = time;
+    }
+
+    public String name() {
+        return name;
+    }
+
+    public StateStore get() {
+        return new MeteredKeyValueStore<>(new RocksDBStore<>(name, serdes), serdes, "rocksdb-state", time).disableLogging();
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableStoreSupplier.java:55]: Could we remove Serdes from the wrapper MeteredKeyValueStore if all internal stores are going to have their own serdes?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableSourceValueGetterSupplier.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        return new KTableSourceValueGetter();
+    }
+
+    private class KTableSourceValueGetter implements KTableValueGetter<K, V> {
+
+        KeyValueStore<K, V> store = null;
+
+        @SuppressWarnings("unchecked")
+        public void init(ProcessorContext context) {
+            if (store == null)
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableSourceValueGetterSupplier.java]: Will this function be triggered more than one time?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java [hunk 8]
```diff
@@ -28,31 +28,37 @@
...
+    protected final Consumer consumer;
     protected final ProcessorStateManager stateMgr;
     protected final Set<TopicPartition> partitions;
     protected ProcessorContext processorContext;
 
     protected AbstractTask(TaskId id,
-                           Consumer<byte[], byte[]> restoreConsumer,
+                           Collection<TopicPartition> partitions,
                            ProcessorTopology topology,
+                           Consumer<byte[], byte[]> consumer,
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:48]: Maybe rename consumer to streamConsumer?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 9]
```diff
@@ -253,14 +282,14 @@ public void close(Map<TopicPartition, Long> ackedOffsets) throws IOException {
...
-                    }
-
                     if (offset != null) {
                         // store the last offset + 1 (the log position after restoration)
                         checkpointOffsets.put(part, offset + 1);
+                    } else {
+                        // if no record was produced. we need to check the restored offset.
+                        offset = restoredOffsets.get(part);
+                        if (offset != null)
+                            checkpointOffsets.put(part, offset);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:292]: Does this change affect the logic?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 10]
```diff
@@ -510,8 +545,6 @@ private void addStreamTasks(Collection<TopicPartition> assignment) {
                 throw e;
             }
         }
-
-        lastClean = time.milliseconds();
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:514]: Is this intentional?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 11]
```diff
@@ -94,6 +96,10 @@
     private long lastCommit;
     private long recordsProcessed;
 
+    private final Set<TopicPartition> standbyPartitions;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: Where is this variable used?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableStoreSupplier.java (hunk 6 above)
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableStoreSupplier.java:55]: In-memory stores do not have own Serdes.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java (hunk 9 above)
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:292]: Yes. We shouldn't do +1 when an offset is from checkpointedOffsets,.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java (hunk 4 above)
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:37]: done

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableSourceValueGetterSupplier.java (hunk 7 above)
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableSourceValueGetterSupplier.java]: no. removed "if".

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 10 above)
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:514]: Yes. It is done in onPartitionAssigned().

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 11 above)
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: It's not used. Removed.

**Comment by guozhangwang** [comment]
LGTM. Jenkins failures are not relevant and have been tracked already.

**Pull request closed without merging**